### PR TITLE
Replaces ReadAccountMapEntry in read_index_for_accessor_or_load_slow()

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

See https://github.com/solana-labs/solana/issues/34786 for background.

We want to limit the use of `ReadAccountMapEntry`, from AccountsIndex, everywhere. Ultimately removing it once there are no more uses.

When reading the index for an account accessor, we currently use `ReadAccountMapEntry` via `AccountsIndex::get()`. But we can do it other ways without `ReadAccountMapEntry`.

#### Summary of Changes

Adds `get_with_and_then()` and uses it in `read_index_for_accessor_or_load_slow()`

### CODE REVIEW

**Review by HaoranYi — COMMENTED** [review]
On code wise, the change in the PR looks good to me. About code comments, it looks like there may be some delicate interaction with "retry". I will defer to jeff to decide how to deal with these comments.

File: accounts-db/src/accounts_index.rs [hunk 1]
```diff
@@ -1149,6 +1149,26 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> AccountsIndex<T, U> {
         self.get_bin(pubkey).get_internal(pubkey, callback)
     }
 
+    /// Gets the index's entry for `pubkey`, with `ancestors` and `max_root`,
+    /// and applies `callback` to it
+    pub fn get_with_and_then<R>(
```
Comment (jeffwashington) [inline accounts-db/src/accounts_index.rs:1154]: `pub(crate)`?

Comment (jeffwashington) [inline accounts-db/src/accounts_index.rs:1154]: I imagine we need tests for this.

**Review by HaoranYi — APPROVED** [review]

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -5077,36 +5077,18 @@ impl AccountsDb {
...
-            None
-        };
-
-        Some((slot, storage_location, some_from_slow_path))
-        // `lock` is dropped here rather pretty quickly with clone_in_lock = false,
-        // so the entry could be raced for mutation by other subsystems,
-        // before we actually provision an account data for caller's use from now on.
-        // This is traded for less contention and resultant performance, introducing fair amount of
-        // delicate handling in retry_to_get_account_accessor() below ;)
-        // you're warned!
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:5118]: it is hard to fit them in. They are outdated in a few ways. `clone_in_lock` is straightforward to understand. I'm ok with removing the comments. The code is simpler.

**Review by jeffwashington — APPROVED** [review]
lgtm

**Pull request merged**